include/ttmlir/Dialect/TTNN/IR/TTNNOpsAttrs.td

```diff
@@ -81,9 +81,9 @@ def TTNN_MemoryConfigAttr : TTNN_Attr<"MemoryConfig", "memory_config"> {
     TTNN memory config attribute
   }];
 
-  let parameters = (ins AttrParameter<"TensorMemoryLayoutAttr", "">:$tensorMemoryLayout,
-                        AttrParameter<"BufferTypeAttr", "">:$bufferType,
-                        AttrParameter<"ShardSpecAttr", "">:$shardSpec);
+  let parameters = (ins AttrParameter<"BufferTypeAttr", "">:$bufferType,
+                        AttrParameter<"ShardSpecAttr", "">:$shardSpec,
+                        OptionalParameter<"TensorMemoryLayoutAttr">:$tensorMemoryLayout);
```

Comment:
Can you revert the ordering? `TensorMemoryLayoutAttr` should be first.

Is it possible to keep the `TensorMemoryLayoutAttr` non-optional? We really shouldn't be creating a `MemoryConfigAttr` if `TensorMemoryLayoutAttr` doesn't exist. I'm fine with keeping it optional for now, until we iron out the quirks in all the places, in which case it would be good to track this in an issue, together with a comment here in code.

Reply:
I did at first... But mlir kept giving me runtime error when trying to parse i.r which looks like this

`memory_config = {<system_memory>, <1x1>}`

I changed assembly format to this:
```
 let assemblyFormat = "`<` ($tensorMemoryLayout^ `,`)? $bufferType `,` $shardSpec `>`";
```

I thought that this was because optionalparams should be last, but I found some examples in llvm where they are in the middle/beginning but I couldn't find some explanation what was wrong with my definition. I will try to find out what is the issue since it's still bugging me, but yea this will be reverted once we stop using to_layout as composite op...

Reply:
Huh, very interesting. If you're in the office tmr, I'd love to take a look with you!

Reply:
It looks like for some reason table gen generates invalid parser for this attribute. From below looks like TensorMemoryLayoutAttr is required :D

```
::mlir::Attribute MemoryConfigAttr::parse(::mlir::AsmParser &odsParser, ::mlir::Type odsType) {
  ::mlir::Builder odsBuilder(odsParser.getContext());
  ::llvm::SMLoc odsLoc = odsParser.getCurrentLocation();
  (void) odsLoc;
  ::mlir::FailureOr<TensorMemoryLayoutAttr> _result_tensorMemoryLayout;
  ::mlir::FailureOr<BufferTypeAttr> _result_bufferType;
  ::mlir::FailureOr<ShardSpecAttr> _result_shardSpec;
  // Parse literal '<'
  if (odsParser.parseLess()) return {};

  // Parse variable 'tensorMemoryLayout'
  _result_tensorMemoryLayout = ::mlir::FieldParser<TensorMemoryLayoutAttr>::parse(odsParser);
  if (::mlir::failed(_result_tensorMemoryLayout)) {
    odsParser.emitError(odsParser.getCurrentLocation(), "failed to parse TTNN_MemoryConfigAttr parameter 'tensorMemoryLayout' which is to be a `TensorMemoryLayoutAttr`");
    return {};
  }
  if (!((::mlir::succeeded(_result_tensorMemoryLayout) && *_result_tensorMemoryLayout))) {
  } else {
    // Parse literal ','
    if (odsParser.parseComma()) return {};
  }

  // Parse variable 'bufferType'
  _result_bufferType = ::mlir::FieldParser<BufferTypeAttr>::parse(odsParser);
  if (::mlir::failed(_result_bufferType)) {
    odsParser.emitError(odsParser.getCurrentLocation(), "failed to parse TTNN_MemoryConfigAttr parameter 'bufferType' which is to be a `BufferTypeAttr`");
    return {};
  }
  // Parse literal ','
  if (odsParser.parseComma()) return {};

  // Parse variable 'shardSpec'
  _result_shardSpec = ::mlir::FieldParser<ShardSpecAttr>::parse(odsParser);
  if (::mlir::failed(_result_shardSpec)) {
    odsParser.emitError(odsParser.getCurrentLocation(), "failed to parse TTNN_MemoryConfigAttr parameter 'shardSpec' which is to be a `ShardSpecAttr`");
    return {};
  }
  // Parse literal '>'
  if (odsParser.parseGreater()) return {};
  assert(::mlir::succeeded(_result_bufferType));
  assert(::mlir::succeeded(_result_shardSpec));
  return MemoryConfigAttr::get(odsParser.getContext(),
      TensorMemoryLayoutAttr((_result_tensorMemoryLayout.value_or(TensorMemoryLayoutAttr()))),
      BufferTypeAttr((*_result_bufferType)),
      ShardSpecAttr((*_result_shardSpec)));
}
```

  let assemblyFormat = "`<` ($tensorMemoryLayout^ `,`)? $bufferType `,` $shardSpec `>`";

Reply:
Okay found some way to make it work. You can set assembly format to
`let assemblyFormat = "`<` struct(params) `>`";`

This will create parser which will require you to specify each parameter with keyword:
`memory_config = #ttnn.memory_config<tensorMemoryLayout = <interleaved>, bufferType = <dram>, shardSpec = <<64x128>>`

I couldn't make it work using optional group like above...

Reply:
>It looks like for some reason table gen generates invalid parser for this attribute. From below looks like TensorMemoryLayoutAttr is required :D

Oh that's interesting haha

Tried to google a little, maaaaybe [this](https://github.com/llvm/llvm-project/commit/07486395d2d05c9c567994456774cafdcc1611d0#diff-53387658ea5255ba7a9a1caa6ab51d32650851b80686bd7cf2320cb9437ae901R489-R520) helps? At the bottom of the section, it has an example assembly where default value is provided in case the attr is null.

It would look like this (if I properly mapped it):
```
 let assemblyFormat = "`<` (($tensorMemoryLayout^):(`None`))?`,` $bufferType `,` $shardSpec `>`";
```

Reply:
That won't work. It will try to parse tensorMemoryLayout first. If it doesn't exits it will try to parse None and if None is not present it will fail.

Reply:
Keep in mind that it's
```
`None`
```
and not
```
None
```
so it would be treated as a literal.

Reply:
I basically copy pasted your example, but regardless if we are to use it we would have to update our IR and set None on all places where TensorMemoryLayout doesn't exits.